connection.c

```diff
@@ -127,29 +128,67 @@ void ConnAsyncTerminate(Connection *conn)
     conn->flags |= CONN_TERMINATING;
 }
 
-/* Connect callback (hiredis).
- *
- * This callback will always be called as a result of a prior redisAsyncConnect()
- * from handleResolved(), with state CONN_CONNECTING.
- *
- * Depending on status, state transitions to CONN_CONNECTED or CONN_CONNECT_ERROR.
- * User callback is called in both cases.
- */
+static void connectionSuccess(Connection *conn)
+{
+    conn->state = CONN_CONNECTED;
+    conn->connect_oks++;
+    conn->last_connected_time = RedisModule_Milliseconds();
 
-static void handleConnected(const redisAsyncContext *c, int status)
+    /* If connection was flagged for termination between connection attempt
+     * and now, we don't call the connect callback.
+     */
+    /* If we're terminating, abort now */
+    if (conn->flags & CONN_TERMINATING) {
+        return;
+    }
+
+    /* Call callback even when failed */
+    if (conn->connect_callback) {
+        conn->connect_callback(conn);
+    }
+
+}
+
+static void connectionFailure(Connection *conn)
 {
-    Connection *conn = (Connection *) c->data;
+    conn->state = CONN_CONNECT_ERROR;
+    conn->rc = NULL;
+    conn->connect_errors++;
 
-    CONN_TRACE(conn, "handleConnected: status=%d", status);
+    /* If connection was flagged for termination between connection attempt
+     * and now, we don't call the connect callback.
+     */
+    /* If we're terminating, abort now */
+    if (conn->flags & CONN_TERMINATING) {
+        return;
+    }
 
-    if (status == REDIS_OK) {
+    /* Call callback even when failed */
+    if (conn->connect_callback) {
+        conn->connect_callback(conn);
+    }
+}
+
+static void handleAuth(redisAsyncContext *c, void *r, void *privdata)
+{
+    Connection *conn = privdata;
+
+    redisReply *reply = r;
+    if (!reply) {
+        LOG_ERROR("Redis connection authentication failed: connection died");
+        ConnMarkDisconnected(conn);
+    } else if (reply->type == REDIS_REPLY_ERROR) {
+        LOG_ERROR("Redis connection authentication failed: %s", reply->str);
+        redisAsyncDisconnect(ConnGetRedisCtx(conn));
+        ConnMarkDisconnected(conn);
+    } else if (reply->type != REDIS_REPLY_STATUS || strcmp("OK", reply->str) != 0) {
+        LOG_ERROR("Redis connection authentication failed: unexpected response");
+        redisAsyncDisconnect(ConnGetRedisCtx(conn));
+        ConnMarkDisconnected(conn);
+    } else {
         conn->state = CONN_CONNECTED;
```

Comment:
Isn't this back to connectionSuccess()?

Reply:
sort of, unsure about asthetics, the issue is that it falls through to after the else to shared code for both success/failure.  pondering